### ISSUE TITLE
Garbage collection - freeing up GPU memory

### ISSUE THREAD

**Issue by MurrayData** [rapidsai/cudf#339]
This may be a question about how Rapids CUDF handles garbage collection. I've been seeing strange results.

I have a multistage process where I need to apply a multistep operations on a dataframe, so creates intermediate dataframes that I no longer need after the stage is complete.

So after I start with nvidia-smi reporting 579MiB in use

So I create a DataFrame gdf1 randomly filled and now have 4361MiB in use

I then apply an apply_rows function creates an additional four columns

gdf2 = gdf1.applyrows(.....

It then shows 10595MiB in use

If I then use the command: def gdf1, it stays the same and the RAM isn't freed.

If I then use the command: del gdf2, it goes down to 6509MiB

If I then recreate gdf1 as before, it shows 10085MiB in use

and recreate gdf2 with apply_rows again it shows 10355MiB in use

but then when I delete gdf2 'del gdf2' then gdf1 'del gdf1', in that order, it drops to 2453MiB in use.

It would appear that it is ok if the deletions are done in reverse order of creation, but either overwriting, by reassignment, or deleting in any other order, doesn't free up all the GPU memory.

if I follow steps as above to recreate gdf1, it shows  4361MiB

but then if I reassign the apply_rows:

gdf1 = gdf1.applyrows(.....

it shows 10595MiB in use again, but the original gdf1 is lost but it's memory hasn't been freed.

if I then delete gdf1 'del gdf1', then it shows 6509MiB, so not all the RAM is being used

I am trying to understand how memory allocation of DataFrames works, so I can make more efficient use of GPU RAM.

P.S. Now I have found I can call Numba JIT CUDA functions, within an apply_rows kernel, I can possibly do all the steps without the need for intermediate DataFrames

**Comment by kkraus14** [comment]
@harrism Does RMM handle the garbage collection by default, or with the standard backend does Numba still handle garbage collection?

**Comment by harrism** [comment]
Numba handles the garbage collection of dataframes in all cases. From the looks of it @MurrayData is using the current default (no pool allocator) but even with the pool allocator RMM just creates a finalizer which is passed to Numba.

**Comment by harrism** [comment]
@MurrayData are you running out of memory at any point?

**Comment by MurrayData** [comment]
@harrism yes I have, several times, when reassigning result to dataframe of same name. I am now bejng careful how I use and release memory. I am using the Docker container. Restarting the kernel does clewr the RAM.

**Comment by kkraus14** [comment]
@MurrayData do you have a snippet that I could use to reproduce the issue to investigate?
